x/programs/cmd/simulator/src/lib.rs

```diff
@@ -271,58 +299,82 @@ where
         .map_err(|e| format!("failed to wait for child: {e}").into())
 }
 
-fn run_steps<P, T>(path: P, plan: &Plan) -> Result<Vec<T>, Box<dyn Error>>
+fn run_steps<T>(
+    stdin: &mut ChildStdin,
+    stdout: &mut ChildStdout,
+    plan: &Plan,
+) -> Result<Vec<T>, Box<dyn Error>>
 where
     T: serde::de::DeserializeOwned + serde::Serialize,
-    P: AsRef<OsStr>,
 {
-    let output = cmd_output(path, plan)?;
     let mut items: Vec<T> = Vec::new();
-
-    if !output.status.success() {
-        println!("stderr");
-        for line in String::from_utf8_lossy(&output.stderr).lines() {
-            println!("{line}");
-        }
-        println!("stdout");
-        for line in String::from_utf8_lossy(&output.stdout).lines() {
-            println!("{line}");
-        }
-        return Err(String::from_utf8(output.stdout)?.into());
+    for step in &plan.steps {
+        let resp = run_step::<T>(
+            stdin,
+            stdout,
+            &Plan {
+                caller_key: plan.caller_key.clone(),
+                steps: vec![step.clone()],
+            },
```

Comment:
Something seems off here. I think you moved the caller-key, right? I think you could abstract this a little better.

Reply:
Yes, the idea is to be able to change the caller from step to step, while still being able to define a plan with the same caller for each step in it. It is now possible to also define a plan with caller A and another one with caller B in the same simulation. I have used a `Step` here instead now.